src/Disks/DiskLocal.cpp

```diff
@@ -317,64 +367,160 @@ SyncGuardPtr DiskLocal::getDirectorySyncGuard(const String & path) const
     return std::make_unique<LocalDirectorySyncGuard>(fs::path(disk_path) / path);
 }
 
-DiskPtr DiskLocalReservation::getDisk(size_t i) const
+bool DiskLocal::setupAndCheck()
 {
-    if (i != 0)
+    try
     {
-        throw Exception("Can't use i != 0 with single disk reservation", ErrorCodes::INCORRECT_DISK_INDEX);
+        fs::create_directories(disk_path);
+    }
+    catch (...)
+    {
+        LOG_WARNING(logger, "Cannot create the directory of disk {} ({}).", name, disk_path);
+        return false;
     }
-    return disk;
-}
-
-void DiskLocalReservation::update(UInt64 new_size)
-{
-    std::lock_guard lock(DiskLocal::reservation_mutex);
-    disk->reserved_bytes -= size;
-    size = new_size;
-    disk->reserved_bytes += size;
-}
-
 
-DiskLocalReservation::~DiskLocalReservation()
-{
     try
     {
-        std::lock_guard lock(DiskLocal::reservation_mutex);
-        if (disk->reserved_bytes < size)
+        if (!FS::canRead(disk_path) || !FS::canWrite(disk_path))
         {
-            disk->reserved_bytes = 0;
-            LOG_ERROR(disk->log, "Unbalanced reservations size for disk '{}'.", disk->getName());
+            LOG_WARNING(logger, "There is no RW access to disk {} ({}).", name, disk_path);
+            return false;
         }
-        else
+    }
+    catch (...)
+    {
+        tryLogCurrentException("DiskLocal");
+        return false;
+    }
+
+    String file_path = disk_path + ".check";
+    const char * fpath = file_path.c_str();
+    char errmsg[64];
+    if (::access(fpath, F_OK) == 0)
+    {
+        if (::remove(fpath) != 0)
         {
-            disk->reserved_bytes -= size;
+            LOG_WARNING(logger, "Fail to delete test file. path = {}, errno = {}, err = {}.", fpath, errno, strerror_r(errno, errmsg, 64));
+            return false;
         }
+    }
+    else
+    {
+        if (errno != ENOENT)
+        {
+            LOG_WARNING(logger, "Fail to access test file. path = {}, errno = {}, err = {}.", fpath, errno, strerror_r(errno, errmsg, 64));
+            return false;
+        }
+    }
+    int fd;
+    if ((fd = ::open(fpath, O_RDWR | O_CREAT | O_SYNC, S_IRUSR | S_IWUSR)) < 0)
+    {
+        LOG_WARNING(logger, "Fail to create test file. path={}.", fpath);
+        return false;
+    }
 
-        if (disk->reservation_count == 0)
-            LOG_ERROR(disk->log, "Unbalanced reservation count for disk '{}'.", disk->getName());
-        else
-            --disk->reservation_count;
+    bool read_write_ok = true;
+    if (disk_checker)
+    {
+        auto rw_test = [fd](char * buf, auto && func)
```

Comment:
Using one code for the read and write looks error prone.

And seems that there is error even now, consider the case when the disk is full and you cannot even write 4K (BUF_SIZE), in this case you will mark the disk as broken, while you should not.

Reply:
I've added `errno==ENOSPC` check.

Reply:
If the disk is full - it is non-writable. The same if filesystem is mounted in readonly after error. Why these cases are different?

Reply:
And if disk is non-writeable for whatever reason, we don't have to repair parts to another disk.

Reply:
>If the disk is full - it is non-writable

It can be temporary full, i.e. due to some temporary files (that can be created by ClickHouse itself)